### ISSUE TITLE
Project file disk image problem on 2.2.22

### ISSUE THREAD

**Issue by theister-xan** [GNS3/gns3-server#1925]
We opend a discussion on the community support forum(https://gns3.com/community/support/project-disk-file-problem-on-2-2-22), but were pointed to open an issue here. I suspect the problem is with gns3-server, so i open it here.

We are running a couple of GNS3 servers on ubuntu 20.04. and recently have upgrade to 2.2.22. It seemed to work fine for some time.

After a recent server reboot (due to kernel updates) and therefore service restart we noticed that all projects started after the reboot/upgrade fail to start the nodes.

After further invetigation we found out that the projects started after the restart have broken disk path images for the nodes.

Some Background:
We have a couple of custom Templates (mostly stuff like juniper vMX in various versions and other VNFs) build and imported. Theses templates are listing fine (e.g. in gns3_controllers.conf) and are listing the correct disk images.

For example we have templates for
vmx20.4 with hda and a disk path of ...../20.4/hda.qcow
vmx19.4 with hda and a disk path of ...../19.4/hda.qcow
vmx18.4 with hda and a disk path of ...../18.4/hda.qcow

Using vmx20.4 it installs the correct disk image path into the gns3 project file in the topologies project dir. And the node starts up fine.

Running 2.2.22 and starting any topology the disk images inside the gns3 project file in the project dir get randomly reasigned to $something that is listed in the controllers.conf

e.g. vmx20.4 could end up with vmx 18.4 hda disk files and/or completly different disk files (we have vmx, vsrx and other vendors images in there). The template id in the gns3 project file for any node is still valid and still points to a template with correct disk image assignements in gns3_controller.conf.

Even manually fixing the disk path in the gns3 file does not help as it will be rearranged once a topology is (re)started.

We reverted the Kernel Update (just in case) but this did not help. 

We than reverted back to 2.2.21 (minor side rant here, why are older releases no longer serverd via the PPA repo? We had to downgrade using the apt cache) and the problem is gone for now.

Of course we needed to downgrade the projects to 2.2.21 and remove the link_style attributes in the project_files which was introduced in 2.2.22. 

After fixing the disk assignments in the broken projects they work fine in 2.2.21 and the assignements are not changed when restarting the gns service or the project. For now we have pinned the version to 2.2.21.

Any hints on what could be the problem or how to troubeshoot this any further?

We even simplified the setup and imported only two custom templates into one of the servers. Even than the disk assignments are getting messed up between these two on project start.

Unfortunately this also cause some data loss as the instances booting with wrong images could missbehave and think that they are a different version and/or not initialized, therefore killing the config that lives on other disks on initial boot.

**Comment by theister-xan** [comment]
Any idea about the problem? Anything we can do to further troubleshoot the issue?

**Comment by grossmj** [comment]
This is very strange. The only thing I found between v2.2.21 and v2.2.22 which could affect images is https://github.com/GNS3/gns3-server/commit/6847e198478d4c53026162f590cdf991199172da#diff-1668d005363a9a4ba3a491125d50d390a9512632d35c6778e2ef195922b60c6c

**Comment by grossmj** [comment]
@theister-xan There is something I don't understand, why do you need to store image outside the default GNS3 images directory?

**Comment by theister-xan** [comment]
@grossmj I am not quite sure what you mean by "outside of the default GNS3 image directory". Our GNS3 install lives in /srv/gns3 and the images live in a hierarchy below /srv/gns3/images. Under that images folder we have built a hierarchy for various vendors versions etc. The config file specifies that image path.

e.g.

```
ls -lha /srv/gns3/images/QEMU/
a10/       cisco/     fortinet/  microsoft/ nvidia/    _special/
arista/    extreme/   juniper/   netscout/  paloalto/
```

and versions

```
ls -lha /srv/gns3/images/QEMU/juniper/vmx/
16.1R7-S7/    17.3R3-S7.2/  18.4R2-S3/    19.2R1-S1.4/  20.1R1-S2/    21.2R1.10/
16.2R2-S10/   17.4R2.4/     18.4R2-S4.10/ 19.3R1.8/     20.2R1.10/
17.3R3-S3.3/  17.4R2-S6/    18.4R3-S4.2/  19.3R2.9/     20.3R1-S1.2/
17.3R3-S5.2/  18.1R1.9/     18.4R3-S7.2/  19.4R1-S1.2/  20.4R1.12/
17.3R3-S6.3/  18.3R1.9/     19.1R1-S3/    19.4R3-S2.2/  21.1R1.11/
```

Having them all in one directory would be quite cumbersome. They are on boarded automatically by one our our scripts that creates the images, copies the file to the locations and also installs/creates the templates.

Would the "non standard directory" be a problem? All locations that we are aware off work on path definitions for the images and should work regardless of where the images are stored (and the do up to including 2.2.21)? The user that gns3 runs under (gns3) has full access to all the image directories and the /srv/gns3 directory infra as well as the controller and config file.

Is there anything we can do to further troubleshoot the issue? 
Right now all of our "production" installs run on 2.2.21 and the version is pinned in apt. But we can spin up a test vm using the exact same setup with 2.2.22 pretty easily. During troubleshooting we saw, that even with two images registered in the template store (and being show with correct path in e.g. gns3_controllers.conf) the image path definitions in the project files would get shuffled on e.g. gns3 service/project restart. We could than provide the gns3 config, the gns3 controller config file, sample project file and the layout of the  images directory if that helps?

**Comment by grossmj** [comment]
@theister-xan sorry for the delay. How do you install GNS3? (ppa or pip or github?)

Also, do you run gns3 with the `--local` parameter?

```
python3 -m gns3server --local
2021-08-24 21:39:20 INFO run.py:219 GNS3 server version 2.2.24dev1-27bcb605
2021-08-24 21:39:20 INFO run.py:221 Copyright (c) 2007-2021 GNS3 Technologies Inc.
2021-08-24 21:39:20 INFO run.py:224 Config file /home/grossmj/.config/GNS3/2.2/gns3_server.conf loaded
2021-08-24 21:39:20 WARNING run.py:230 Local mode is enabled. Beware, clients will have full control on your filesystem
```

Thanks.

**Comment by theister-xan** [comment]
@grossmj it is/was installed via apt/ppa from http://ppa.launchpad.net/gns3/ppa/ubuntu

the server/daemon itself is started using a systemd unit

```
Type=forking
User=gns3
Group=gns3
PermissionsStartOnly=true
ExecStart=/usr/bin/gns3server --log /var/log/gns3/gns3.log  --pid /var/run/gns3/gns3.pid --daemon --config /etc/gns3-server.conf
ExecReload=/bin/kill -s HUP $MAINPID
Restart=on-abort
PIDFile=/var/run/gns3/gns3.pid

```

We are running multiple physical Servers each running a GNS3 instance and access to these instances is done either using the Web-UI or the Client Software installed on the Clients

**Comment by theister-xan** [comment]
What would that change, and why do we think that it is related to the specific issue?

I need to setup a "broken" instance later this week. And hence older version are not available via ppa i would need to check whether the "broken" version still is around in some apt archive on one of the other servers as we are now on 2.2.23 in the ppa/repo.

**Comment by grossmj** [comment]
> What would that change, and why do we think that it is related to the specific issue?

The `--local` param allows the server to access any file outside the default images directory. I am just trying to guess where the problem might be, I think it may be because of these changes https://github.com/GNS3/gns3-server/commit/6294ad9e76ba0984dd649022b13e0a2aad2b7637 we have done between v2.2.21 and v2.2.22

> I need to setup a "broken" instance later this week. And hence older version are not available via ppa i would need to check whether the "broken" version still is around in some apt archive on one of the other servers as we are now on 2.2.23 in the ppa/repo.

Thanks, you may want to try to install directly using Python pip which is more flexible. Let me know if you need any help with that.

**Comment by theister-xan** [comment]
@grossmj 
I managed to put some time into it today. Fresh VM installed via our GNS3 FAI with 2.2.23 from ppa, manually downgraded to 2.2.22. Started the GNS3 Server with --local

`gns3        2820  1.1  0.2 358444 47252 ?        Sl   16:04   0:03 /usr/share/gns3/gns3-server/bin/python /usr/bin/gns3server --log /var/log/gns3/gns3.log --pid /var/run/gns3/gns3.pid --daemon --config /etc/gns3-server.conf --local`

```
...
2021-08-24 16:04:07 INFO run.py:219 GNS3 server version 2.2.22
2021-08-24 16:04:07 INFO run.py:221 Copyright (c) 2007-2021 GNS3 Technologies Inc.
2021-08-24 16:04:07 INFO run.py:224 Config file /etc/gns3-server.conf loaded
2021-08-24 16:04:07 WARNING run.py:230 Local mode is enabled. Beware, clients will have full control on your filesystem
2021-08-24 16:04:07 INFO run.py:237 HTTP authentication is enabled with username 'gns3'
2021-08-24 16:04:07 INFO run.py:243 Running with Python 3.8.5 and has PID 2820
2021-08-24 16:04:07 INFO run.py:79 Current locale is en_US.UTF-8
2021-08-24 16:04:07 INFO web_server.py:318 Starting server on 0.0.0.0:3080
2021-08-24 16:04:07 INFO __init__.py:63 Load controller configuration file /etc/gns3_controller.conf
2021-08-24 16:04:07 INFO __init__.py:67 Controller is starting
....
```

When on boarding one image/template into gns3 all is fine:
```
 grep /"vmx" /etc/gns3_controller.conf
            "hda_disk_image": "juniper/vmx/18.4R3-S7.2/re/hda.qcow2",
            "hdb_disk_image": "juniper/vmx/18.4R3-S7.2/re/hdb.qcow2",
            "hdc_disk_image": "juniper/vmx/18.4R3-S7.2/re/hdc_re0.qcow2",
            "hda_disk_image": "juniper/vmx/18.4R3-S7.2/re/hda.qcow2",
            "hdb_disk_image": "juniper/vmx/18.4R3-S7.2/re/hdb.qcow2",
            "hdc_disk_image": "juniper/vmx/18.4R3-S7.2/re/hdc_re1.qcow2",
            "hda_disk_image": "juniper/vmx/18.4R3-S7.2/fpc/hda.qcow2",
```

```
/srv/gns3/projects/04e20299-3453-46d7-b5c6-e6d92c474deb# grep "vmx/" test-topo1.gns3
                    "hda_disk_image": "juniper/vmx/18.4R3-S7.2/fpc/hda.qcow2",
                    "hda_disk_image": "juniper/vmx/18.4R3-S7.2/re/hda.qcow2",
                    "hdb_disk_image": "juniper/vmx/18.4R3-S7.2/re/hdb.qcow2",
                    "hdc_disk_image": "juniper/vmx/18.4R3-S7.2/re/hdc_re0.qcow2",
```

On boarding a second image/template and it looks like this:
```
grep /"vmx" /etc/gns3_controller.conf
            "hda_disk_image": "juniper/vmx/18.4R3-S7.2/re/hda.qcow2",
            "hdb_disk_image": "juniper/vmx/18.4R3-S7.2/re/hdb.qcow2",
            "hdc_disk_image": "juniper/vmx/18.4R3-S7.2/re/hdc_re0.qcow2",
            "hda_disk_image": "juniper/vmx/18.4R3-S7.2/re/hda.qcow2",
            "hdb_disk_image": "juniper/vmx/18.4R3-S7.2/re/hdb.qcow2",
            "hdc_disk_image": "juniper/vmx/18.4R3-S7.2/re/hdc_re1.qcow2",
            "hda_disk_image": "juniper/vmx/18.4R3-S7.2/fpc/hda.qcow2",
            "hda_disk_image": "juniper/vmx/21.1R1.11/re/hda.qcow2",
            "hdb_disk_image": "juniper/vmx/21.1R1.11/re/hdb.qcow2",
            "hdc_disk_image": "juniper/vmx/21.1R1.11/re/hdc_re0.qcow2",
            "hda_disk_image": "juniper/vmx/21.1R1.11/re/hda.qcow2",
            "hdb_disk_image": "juniper/vmx/21.1R1.11/re/hdb.qcow2",
            "hdc_disk_image": "juniper/vmx/21.1R1.11/re/hdc_re1.qcow2",
            "hda_disk_image": "juniper/vmx/21.1R1.11/fpc/hda.qcow2",
```
we than restart the gns3 service:
`systemctl restart gns3.service`

If the old topology has not yet been (re)started it still looks like this:
```
grep "vmx/" test-topo1.gns3
                    "hda_disk_image": "juniper/vmx/18.4R3-S7.2/fpc/hda.qcow2",
                    "hda_disk_image": "juniper/vmx/18.4R3-S7.2/re/hda.qcow2",
                    "hdb_disk_image": "juniper/vmx/18.4R3-S7.2/re/hdb.qcow2",
                    "hdc_disk_image": "juniper/vmx/18.4R3-S7.2/re/hdc_re0.qcow2",
```

Once we (re)start the topology (e.g. Web-UI) the topology file looks like this:
```
grep "vmx/" test-topo1.gns3
                    "hda_disk_image": "juniper/vmx/21.1R1.11/fpc/hda.qcow2",
                    "hda_disk_image": "juniper/vmx/21.1R1.11/re/hda.qcow2",
                    "hdb_disk_image": "juniper/vmx/21.1R1.11/re/hdb.qcow2",
                    "hdc_disk_image": "juniper/vmx/21.1R1.11/re/hdc_re0.qcow2",
```
and will no longer work.

**Comment by grossmj** [comment]
Thanks, I am going to try to replicate this issue on my side.

One more question, do you have any errors or warnings in `/var/log/gns3/gns3.log`?

**Comment by theister-xan** [comment]
not really the are some warning because i did not import/add the images/svgs to the test instance:
2021-08-24 16:04:25 WARNING symbols.py:135 Could not retrieve symbol 'juniper-vmx.svg', returning default symbol...

obviously there is the "local" Warning:
2021-08-24 16:04:07 WARNING run.py:230 Local mode is enabled. Beware, clients will have full control on your filesystem

Messages on restarting the server:
2021-08-24 16:04:04 WARNING web_server.py:170 Server has got signal SIGTERM, exiting...

the only other thing in there is:
2021-08-24 16:02:09 WARNING compute.py:367 Cannot connect to compute 'local': Cannot connect to host 127.0.0.1:3080 ssl:default [Connect call failed ('127.0.0.1', 3080)]
Which seems to be reported on each report if a client was/is connected

**Comment by grossmj** [comment]
Can you post the content of gns3_controller.conf? Use https://pastebin.com/ if its too big for GitHub

**Comment by theister-xan** [comment]
gns3_controller.conf: https://pastebin.com/71sfnemb

i also added the "broken" topology file after the reload in case you need it: https://pastebin.com/dAD1pkyq

You can see, that the template IDs are still valid, but the disk image path has changed.

**Comment by grossmj** [comment]
@theister-xan looks like I found the sneaky bug. The issue should be fixed now, we will release version 2.2.24 this week. Thanks again for reporting the problem and helping with the debugging :+1:

**Comment by grossmj** [comment]
@theister-xan  have you had a chance to test? just to confirm we really fixed the problem ;) thanks 👍

**Comment by theister-xan** [comment]
@grossmj 
Sorry, I was not closely monitoring the release of the fixed version. I upgraded the test VM we used for the problem replication and it seems OK after the upgrade. A new topology with two different VNF versions keeps the project file intact even after service restarts.
A more in depth test on the production machines will probably take a couple of weeks as its still vacation time around here :)
